src/github.com/getlantern/lantern-android/client/conn.go

```diff
@@ -0,0 +1,63 @@
+package client
+
+import (
+	"errors"
+	"net"
+)
+
+var (
+	// ErrCouldNotCreateListener is returned when NewListener fails.
+	ErrCouldNotCreateListener = errors.New(`Could not create new listener.`)
+	// ErrClosed is returned when the server is closed.
+	ErrClosed = errors.New(`Server was manually closed.`)
+)
+
+// Listener is a wrapper around net.TCPListener that attempts to provide a
+// Stop() function.
+type Listener struct {
+	*net.TCPListener
+	closed chan bool
+}
+
+// NewListener creates a wrapper around TCPListener
+func NewListener(addr string) (wrap *Listener, err error) {
+	var li net.Listener
+	var tli *net.TCPListener
+
+	var ok bool
+
+	if li, err = net.Listen("tcp", addr); err != nil {
+		return nil, err
+	}
+
+	if tli, ok = li.(*net.TCPListener); !ok {
+		return nil, ErrCouldNotCreateListener
+	}
+
+	wrap = &Listener{
+		TCPListener: tli,
+		closed:      make(chan bool),
+	}
+
+	return wrap, nil
+}
+
+// Accept returns the next connection to the listener.
+func (li *Listener) Accept() (net.Conn, error) {
+	for {
+		select {
+		case <-li.closed:
+			return nil, ErrClosed
+		default:
+			return li.TCPListener.Accept()
+		}
```

Comment:
> Is there a reason you don't just use a regular listener and call its Close() method when you want to stop listening?

No, it was a stupid mistake. I was actually trying to create an error within the `http.Server.Serve()` method, specifically on this line: http://golang.org/src/net/http/server.go#L1728 that stopped the http server, so I thought that I could wrap `li.Accept()` to make it produce an error when I wanted. But `Close()` produced the error I needed already. I could swear I made a test before with a plain net.Listener that didn't work and kept blocking instead.

Sorry, I don't know what was I thinking but it had to be something rather dumb. I removed this thing completely from the source (even though it will remain in my memory for a while, haha).

Thank you very much for pointing this out @oxtoacart :-).